### ISSUE TITLE
[Wiki] StatisticalBehaviour - seed ≠ deterministic?

### ISSUE THREAD

**Issue by Fylwind** [simulationcraft/simc#9818]
**Documentation page:** https://github.com/simulationcraft/simc/wiki/StatisticalBehaviour
**Source code version:** 9cabab08cf28e56a744071e300daf766850846ed
**Build flags:** `CMAKE_BUILD_TYPE=Release`

The documentation says that `deterministic=1` is equivalent to the option `seed=31459`:

> - **deterministic** (scope:global, default:0), when different from zero, will seed our rng packages with an arbitrary, hard-coded, value (31459, incremented for every thread). Equal to the option seed=31459.

However, this does not seem to be true. I tested these two profiles and got two different results:

```
threads=1
iterations=1
deterministic=1
input=simc/profiles/TWW1/TWW1_Monk_Windwalker.simc
json2=output1.json
# output.sim.players[0]collected_data.dps.mean = 1638343.89832064
```

```
threads=1
iterations=1
seed=31459
input=simc/profiles/TWW1/TWW1_Monk_Windwalker.simc
json2=output2.json
# output.sim.players[0]collected_data.dps.mean = 1650896.256411524
```

Furthermore, it looks like `seed` alone, without `determinism`, fails to prevent nondeterminism when `threads` and `iterations` are both greater than one:

```
threads=2
iterations=3
seed=31459
input=/home/fyl/stuff/simc/profiles/TWW1/TWW1_Monk_Windwalker.simc
json2=output3.json
# output.sim.players[0]collected_data.dps.mean = randomly either 1653143.593394113 or 1655614.598596016
```

**Comment by vituscze** [comment]
> However, this does not seem to be true. I tested these two profiles and got two different results:

You are correct that the documentation is missing something. `deterministic=1` causes the RNG to reseed itself after every iteration (otherwise the state of the RNG might not correspond to any seed). As a side effect, the first iteration won't start with the seed 31459 but rather the next 64 bits the RNG produces.

> Furthermore, it looks like seed alone, without determinism, fails to prevent nondeterminism when threads and iterations are both greater than one:

You want to run the sim with `strict_work_queue=1` so that each thread gets a predictable number of iterations to run. But unless you *really* need to have a deterministic multi-threaded sim, this is a bad idea as it might result in threads idling near the end of the sim.

**Comment by Fylwind** [comment]
Thanks for the quick response!

> You want to run the sim with `strict_work_queue=1` so that each thread gets a predictable number of iterations to run.

What if each thread pulled from a global thread-safe RNG, once per iteration, and then used that value to seed a thread-local RNG for that iteration? That could allow more rebalancing between threads and allow the result to be the same regardless of thread count.

**Comment by vituscze** [comment]
Sure, I'm not saying that the current approach cannot be improved. It's a question of whether it's adding anything useful. The `deterministic` and `seed` (and also `debug_seed`) options exist to be able to replay single iterations for debugging purposes (whether it's simc crashes or poorly performing APL), anything else is an afterthought.
